File: mvvm-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsViewModel.java
package com.wasisto.githubuserfinder.android.ui.userdetails;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.wasisto.githubuserfinder.android.ui.common.Event;
import com.wasisto.githubuserfinder.common.Logger;
import com.wasisto.githubuserfinder.domain.models.User;
import com.wasisto.githubuserfinder.domain.usecases.GetSearchHistoryUseCase;
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase;
import com.wasisto.githubuserfinder.domain.usecases.SearchUsersUseCase;

import java.util.concurrent.ExecutorService;

public class UserDetailsViewModel extends ViewModel {

    private GetUserUseCase getUserUseCase;

    private Logger logger;

    private ExecutorService ioExecutorService;

    private ExecutorService mainExecutorService;

    private MutableLiveData<User> user = new MutableLiveData<>();

    private MutableLiveData<Boolean> shouldShowLoadingIndicator = new MutableLiveData<>();

    private MutableLiveData<Event<String>> openBrowserEvent = new MutableLiveData<>();

    private MutableLiveData<Event<Void>> showErrorToastEvent = new MutableLiveData<>();

    private MutableLiveData<Event<Void>> closeActivityEvent = new MutableLiveData<>();

    public UserDetailsViewModel(GetUserUseCase getUserUseCase, Logger logger, ExecutorService ioExecutorService,
                                ExecutorService mainExecutorService) {
        this.getUserUseCase = getUserUseCase;
        this.logger = logger;
        this.ioExecutorService = ioExecutorService;
        this.mainExecutorService = mainExecutorService;
    }

    public void load(String username) {
        mainExecutorService.execute(() -> {
            shouldShowLoadingIndicator.setValue(true);

            ioExecutorService.execute(() -> {
                try {
                    User retrievedUser = getUserUseCase.execute(username);

                    logger.debug(String.format("retrievedUser: %s", retrievedUser));

                    mainExecutorService.execute(() -> {
                        shouldShowLoadingIndicator.setValue(false);
                        user.setValue(retrievedUser);
                    });
                } catch (Exception e) {
                    logger.warn("An error occurred while getting user details", e);

                    mainExecutorService.execute(() -> {
                        showErrorToastEvent.setValue(new Event<>(null));
                        closeActivityEvent.setValue(new Event<>(null));
                    });
                }
            });
        });
    }

    public void onBlogClick() {
        if (user.getValue() != null && user.getValue().getBlog() != null) {
            mainExecutorService.execute(() -> {
                openBrowserEvent.setValue(new Event<>(null));
            });
        }
    }

    public LiveData<User> getUser() {
        return user;
    }

    public LiveData<Boolean> shouldShowLoadingIndicator() {
        return shouldShowLoadingIndicator;
    }

    public LiveData<Event<String>> getOpenBrowserEvent() {
        return openBrowserEvent;
    }

    public LiveData<Event<Void>> getShowErrorToastEvent() {
        return showErrorToastEvent;
    }

    public LiveData<Event<Void>> getCloseActivityEvent() {
        return closeActivityEvent;
    }
}

File: mvvm-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/userdetails/UserDetailsViewModel.kt
package com.wasisto.githubuserfinder.android.ui.userdetails

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.wasisto.githubuserfinder.android.ui.common.Event
import com.wasisto.githubuserfinder.common.Logger
import com.wasisto.githubuserfinder.domain.models.User
import com.wasisto.githubuserfinder.domain.usecases.GetUserUseCase
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.launch

class UserDetailsViewModel(
    private val getUserUseCase: GetUserUseCase,
    private val logger: Logger,
    private val ioDispatcher: CoroutineDispatcher,
    private val mainDispatcher: CoroutineDispatcher,
) : ViewModel() {

    val user = MutableLiveData<User>()

    val shouldShowLoadingIndicator = MutableLiveData<Boolean>()

    val openBrowserEvent = MutableLiveData<Event<String>> ()

    val showErrorToastEvent = MutableLiveData<Event<Unit>>()

    val closeActivityEvent = MutableLiveData<Event<Unit>>()

    fun load(username: String) = viewModelScope.launch(mainDispatcher) {
        shouldShowLoadingIndicator.value = true

        viewModelScope.launch(ioDispatcher) {
            try {
                val retrievedUser = getUserUseCase(username)

                logger.debug("retrievedUser: $retrievedUser")

                launch(mainDispatcher) {
                    shouldShowLoadingIndicator.value = false
                    user.value = retrievedUser
                }
            } catch (e: Exception) {
                logger.warn("An error occurred while getting user details", e)

                launch(mainDispatcher) {
                    showErrorToastEvent.value = Event(Unit)
                    closeActivityEvent.value = Event(Unit)
                }
            }
        }
    }

    fun onBlogClick() = viewModelScope.launch(mainDispatcher) {
        user.value?.blog?.let { blog ->
            openBrowserEvent.value = Event(blog)
        }
    }
}